test/test_optim.py

```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
+            # for `gradcheck` to correctly track the state tensors as function inputs
+            # because otherwise it can't unpack the values in the `opt_differentiable_state`
+            # dict
+            p = p.clone()
+            p.grad = grad
+            opt_differentiable_state = {
+                k: v.clone() if isinstance(v, torch.Tensor) else v
+                for k, v in opt_differentiable_state.items()
+            }
+            opt = opt_class([p], **kwargs)
+            opt.state[p].update(opt_differentiable_state)
+            opt.step()
+            return (p,) + tuple(
+                v
+                for v in opt.state[p].values()
+                if isinstance(v, torch.Tensor) and v.requires_grad
+            )
 
-instantiate_device_type_tests(TestOptimRenewed, globals(), allow_mps=True)
+        def _get_optimizer_state(
+            optim_cls: type[Optimizer],
+            params: torch.Tensor,
+            optim_kwargs: dict[str, Any],
+        ) -> dict[str, Any]:
+            params.grad = torch.rand_like(params)
+
+            optim = optim_cls([params], **optim_kwargs)
+            optim.step()
+
+            states = optim.state_dict()["state"]
+            # Accesses state of the first (and only) parameter. When using SGD w/ no momentum, there is no state
+            state = states[0] if 0 in states else states
+            return state
+
+        optim_cls = optim_info.optim_cls
+        if "differentiable" not in optim_info.supported_impls:
+            self.skipTest(f"Differentiable {optim_cls.__name__} not supported")
+
+        for optim_input in optim_info.optim_inputs_func(device=device):
+            mock_params = torch.rand(10, requires_grad=True, device=device, dtype=dtype)
+            mock_optim_kwargs = optim_input.kwargs.copy()
+            optim_state = _get_optimizer_state(
+                optim_cls, mock_params, mock_optim_kwargs
+            )
 
+            optim_kwargs = optim_input.kwargs
+            optim_kwargs.update({"differentiable": True})
+
+            # Cannot be capturable and differentiable
+            if "capturable" in optim_kwargs:
+                optim_kwargs.update({"capturable": False})
+
+            # sgd defaults to foreach on cuda so must explicitly specify not foreach
+            if optim_cls is SGD and device.startswith("cuda"):
+                optim_kwargs.update({"foreach": False})
+
+            params = torch.rand(10, requires_grad=True, device=device, dtype=dtype)
+            grad = torch.rand(10, requires_grad=True, device=device, dtype=dtype)
+            gradcheck(
+                _diff_fn,
+                (
+                    params,
+                    grad,
+                    optim_state,
+                    optim_cls,
+                    optim_kwargs,
+                    *optim_state.values(),
+                ),
+            )
+
+
+instantiate_device_type_tests(TestOptimRenewed, globals(), allow_mps=True)
```

Comment:
Can you include a printout of what test cases run for this? We want to make sure we're strictly increasing coverage.